digiwf-engine/digiwf-engine-rest-service/src/main/java/de/muenchen/oss/digiwf/adapter/out/ldap/LdapAdapter.java

```diff
@@ -93,15 +102,113 @@ private String resolveUserDn(@NonNull final String username) {
+                .where("objectclass").is("user")
+                .and("cn").is(username);
+        List<User> result = super.search(query, (AttributesMapper<User>) attrs ->
+                User.builder().id(username)
+                        .firstName(attrs.get("givenName").get().toString())
+                        .lastName(attrs.get("sn").get().toString())
+                        .email(attrs.get("mail").get().toString())
+                        .build()
+        );
+        if (result.isEmpty()) {
+            log.info("Username {} not found", username);
+            return null;
+        }
+        if (result.size() > 1) {
+            throw new IllegalStateException(String.format("Multiple users found for username '%s'", username));
+        }
+        val user = result.get(0);
+        log.info("Resolved user {} to {}", username, user);
+        return user;
+    }
+
+    @NonNull
+    @Override
+    @Cacheable(LdapCacheConfiguration.GROUPS_MEMBERS)
+    public List<String> getGroupsMembers(@NonNull @NotEmpty final List<Group> groups) {
+        val resolvedGroups = resolveGroups(groups);
+        val resolvedGroupsCn = resolvedGroups.stream()
+                .flatMap(i -> i.subGroups().stream())
+                .map(this::dnToCn)
+                .map(Group::new).toList();
+        log.info("Resolved {} engine groups to {} ldap groups", groups.size(), resolvedGroups.size());
+        // resolve one level of recursion
+        resolvedGroups.addAll(resolveGroups(resolvedGroupsCn));
+        // map groups with users to unique users
+        return resolvedGroups.stream()
+                .flatMap(i -> i.users().stream())
+                .map(this::dnToCn)
+                .distinct().sorted().toList();
+    }
+
+    /**
+     * Resolves a list of groups to subgroups and member users.
+     *
+     * @param groups List of groups to resolve.
+     * @return List of LdapUsers with corresponding subgroups and member users.
+     */
+    private List<LdapGroup> resolveGroups(@NonNull @NotEmpty final List<Group> groups) {
+        log.trace("Resolving groups {}", groups);
+        // build ldap search filter and query
+        val groupNameFilter = new OrFilter();
+        for (Group group : groups) {
+            groupNameFilter.or(new EqualsFilter("cn", group.name()));
+        }
+        val filter = new AndFilter();
+        filter.and(groupNameFilter);
+        filter.and(new EqualsFilter("objectclass", "group"));
+        LdapQuery query = LdapQueryBuilder.query()
+                .base(properties.getGroupBase())
+                .filter(filter);
+        // search and resolve groups
+        return super.search(query, (AttributesMapper<LdapGroup>) attrs -> {
+            val users = new ArrayList<String>(List.of());
+            val childGroups = new ArrayList<String>(List.of());
+            if (attrs.get("member") != null) {
+                attrs.get("member").getAll().asIterator().forEachRemaining((i) -> {
```

Comment:
Kann man ohne Klammern schreiben:
```suggestion
                attrs.get("member").getAll().asIterator().forEachRemaining(i -> {
```